### PR TITLE
OCPBUGS-10875: gateway-service-dns: Set DNS policy appropriately

### PR DESCRIPTION
**Pull request by Miciah**
#### gateway-service-dns: Add test coverage for deletes

* `pkg/operator/controller/gateway-service-dns/controller_test.go` (`Test_Reconcile`): Add labels to DNS records so that `deleteStaleDNSRecordsForGateway` doesn't filter them out.  Verify that objects that are expected to be deleted are indeed deleted.  Add a test case to verify that deletion works properly.
(`fakeClientRecorder`): Add "deleted" field.
(`Create`, `Delete`, `Update`): Log actions.
(`Delete`): Record deleted objects.

#### gateway-service-dns: Set DNS policy appropriately

When creating a DNSRecord CR for a gateway listener, set the DNS management policy to "Unmanaged" if the DNSRecord's domain is outside the cluster's base domain.

Before this change, the operator would attempt to create DNS records for gateway listeners even if they had domains outside of the cluster's base domain zone, resulting in "failed to publish DNS record to zone" errors.

* `pkg/operator/controller/gateway-service-dns/controller.go` (`Reconcile`): Get the dns and infrastructure config objects, and pass them to `ensureDNSRecordsForGateway`.
(`ensureDNSRecordsForGateway`): Add `infraConfig` and `dnsConfig` parameters. Use them and the `ManageDNSForDomain` function to determine the appropriate policy (managed or unmanaged) for each DNS record, and pass that policy to `EnsureDNSRecord`.
* `pkg/resources/dnsrecord/dns.go` (`EnsureDNSRecord`): Add a `dnsPolicy` parameter, and pass it to `desiredDNSRecord`.
(`ManageDNSForDomain`): Ignore trailing dots when comparing the given domain and the cluster's base domain.
* `pkg/resources/dnsrecord/dns_test.go` (`TestManageDNSForDomain`): Add test cases for empty base domain and for various combinations of trailing dots on the domain or base domain.
* `pkg/operator/controller/gateway-service-dns/controller_test.go` (`Test_Reconcile`): Add `dnsConfig` and `infraConfig` to `existingObjects` in test cases.  Add a test case to verify that the DNS management policy is set to "Unmanaged" if the DNS name doesn't belong to the cluster's base domain.

### CODE REVIEW

File: pkg/operator/controller/gateway-service-dns/controller_test.go [hunk 1]
```diff
@@ -96,79 +118,96 @@ func Test_Reconcile(t *testing.T) {
...
+				),
+				svc("example-gateway", gatewayManagedLabel, exampleGatewayLabel, ingHost("lb.example.com")),
+			},
+			reconcileRequest: req("openshift-ingress", "example-gateway"),
+			expectCreate: []client.Object{
+				dnsrecord("example-gateway-957dccf84-wildcard", "*.foo.com", iov1.UnmanagedDNS, exampleGatewayLabel, "lb.example.com"),
+			},
+			expectUpdate: []client.Object{},
+			expectDelete: []client.Object{},
+		},
```
Comment (gcs278) [inline pkg/operator/controller/gateway-service-dns/controller_test.go:210]: Is it worth adding 2 simple test cases: one for `dnsConfig` is null and another for `infraConfig` is null and nothing should be created?

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/c5b6502afa09f81617c0955e10cd1dffed3ccdb1..8b20178b30c5e9516ac68e8c46649ba2b9d4b701 rebases for #949.

**Comment by Miciah** [comment]
Whoops, https://github.com/openshift/cluster-ingress-operator/compare/8b20178b30c5e9516ac68e8c46649ba2b9d4b701..038190ece2d957a26615bbc666ad9164a2e67846 re-adds code from #949 that got inadvertently got dropped in the rebase.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/038190ece2d957a26615bbc666ad9164a2e67846..0eb49fbfeccdb763ad5757025967393ac0f1465c adds test cases suggested in https://github.com/openshift/cluster-ingress-operator/pull/934#discussion_r1242288616.

File: pkg/operator/controller/gateway-service-dns/controller_test.go (hunk 1 above)
Comment (Miciah) [reply pkg/operator/controller/gateway-service-dns/controller_test.go:247]: Done in https://github.com/openshift/cluster-ingress-operator/compare/038190ece2d957a26615bbc666ad9164a2e67846..0eb49fbfeccdb763ad5757025967393ac0f1465c.

**Comment by Miciah** [comment]
I'll rebase again once https://github.com/openshift/cluster-ingress-operator/pull/949/files#r1245950185 is resolved.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/0eb49fbfeccdb763ad5757025967393ac0f1465c..26492dbda956f0542dbcb9abaec20535be13c56a rebases for #956.

File: pkg/resources/dnsrecord/dns.go [hunk 2]
```diff
@@ -229,11 +229,20 @@ func dnsRecordChanged(current, expected *iov1.DNSRecord) (bool, *iov1.DNSRecord)
 // once we know there are no users depending on this.
 // See https://bugzilla.redhat.com/show_bug.cgi?id=2041616
 func ManageDNSForDomain(domain string, status *configv1.PlatformStatus, dnsConfig *configv1.DNS) bool {
-	if len(domain) == 0 {
+	if len(domain) == 0 || len(dnsConfig.Spec.BaseDomain) == 0 {
```
[the remark this answers is not in the document]
Comment (Miciah) [reply pkg/resources/dnsrecord/dns.go:232]: Added in https://github.com/openshift/cluster-ingress-operator/compare/26492dbda956f0542dbcb9abaec20535be13c56a..620c930bbb51728f597d6ae2db46d04a3974de1a.

File: pkg/resources/dnsrecord/dns.go [hunk 3]
```diff
@@ -229,11 +229,20 @@ func dnsRecordChanged(current, expected *iov1.DNSRecord) (bool, *iov1.DNSRecord)
...
 		return false
 	}
 
 	mustContain := "." + dnsConfig.Spec.BaseDomain
+
+	// Ignore any trailing dot for comparison.
```
[the remark this answers is not in the document]
Comment (Miciah) [reply pkg/resources/dnsrecord/dns.go:238]: Added in https://github.com/openshift/cluster-ingress-operator/compare/26492dbda956f0542dbcb9abaec20535be13c56a..620c930bbb51728f597d6ae2db46d04a3974de1a.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/26492dbda956f0542dbcb9abaec20535be13c56a..620c930bbb51728f597d6ae2db46d04a3974de1a adds test cases requested in https://github.com/openshift/cluster-ingress-operator/pull/934#pullrequestreview-1515048144.

**Comment by Miciah** [comment]
[e2e-aws-ovn-serial failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/934/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-ovn-serial/1681385114858164224) on `[sig-scheduling] Multi-AZ Clusters should spread the pods of a service across zones`:
```
{  fail [test/e2e/scheduling/ubernetes_lite.go:178]: Pods were not evenly spread across zones.  0 in one zone and 10 in another zone
Expected
    <int>: 10
to be within 2 of ~
    <int>: 0
Error: exit with code 1
Ginkgo exit error 1: exit with code 1}
```
Search.ci isn't currently loading for me, so I cannot check whether this failure is occurring on other PRs.
/test e2e-aws-ovn-serial

[e2e-hypershift failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/934/pull-ci-openshift-cluster-ingress-operator-master-e2e-hypershift/1681385121208340480).  It isn't clear why.  I see numerous errors in test output, but no actual test failure.  I think the tests just timed out:
```
 {"component":"entrypoint","error":"process timed out","file":"k8s.io/test-infra/prow/entrypoint/run.go:79","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2023-07-18T20:59:02Z"} 
```
@enxebre, do you know how to diagnose these failures?
/test e2e-hypershift

**Comment by gcs278** [comment]
Sorry for the delay, I missed the notifications on this one.

Updates look good, thanks.
/lgtm
/approve

**Comment by Miciah** [comment]
[e2e-gcp-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/934/pull-ci-openshift-cluster-ingress-operator-master-e2e-gcp-operator/1687266371919941632) because must-gather failed.

Also, #968 merged, so we need to rerun all tests.
/test all.

**Comment by Miciah** [comment]
[e2e-aws-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/934/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1687619514184765440) because must-gather failed.
/test e2e-aws-operator

[e2e-aws-ovn-serial failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/934/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-ovn-serial/1687619514235097088) because `[sig-storage] PersistentVolumes-local Stress with local volumes [Serial] should be able to process many pods and reuse local volumes` failed:
```text
{  fail [test/e2e/storage/persistent_volumes-local.go:522]: persistentvolumes "local-pvx9bhx" not found
Error: exit with code 1
Ginkgo exit error 1: exit with code 1}
```
This is a known issue: [OCPBUGS-14930](https://issues.redhat.com/browse/OCPBUGS-14930).
/test e2e-aws-ovn-seria

/hold cancel
